src/common/libflux/rpc.c

```diff
@@ -28,194 +28,259 @@
+        return -1;
+    rpc->rx_msg_consumed = false;
+    rpc->rx_count++;
+    return 0;
+}
+
+static int rpc_sendmsg (flux_rpc_t rpc, int n, const char *topic,
+                        const char *json_str, uint32_t nodeid)
+{
+    zmsg_t *zmsg;
+    int flags = 0;
     int rc = -1;
-    int errnum;
-    const char *json_str;
-    JSON o = NULL;
 
-    if (match.matchtag == FLUX_MATCHTAG_NONE) {
-        errno = EAGAIN;
-        goto done;
-    }
-    if (flux_json_request (h, nodeid, match.matchtag, topic, in) < 0)
+    if (!(zmsg = flux_request_encode (topic, json_str)))
         goto done;
-    if (!(zmsg = flux_recvmsg_match (h, match, NULL, false)))
-        goto done;
-    if (flux_msg_get_errnum (zmsg, &errnum) < 0)
-        goto done;
-    if (errnum != 0) {
-        errno = errnum;
+    if (flux_msg_set_matchtag (zmsg, rpc->m.matchtag + n) < 0)
         goto done;
+    if (nodeid == FLUX_NODEID_UPSTREAM) {
+        flags |= FLUX_MSGFLAG_UPSTREAM;
+        nodeid = flux_rank (rpc->h);
     }
-    if (flux_msg_get_payload_json (zmsg, &json_str) < 0)
+    if (flux_msg_set_nodeid (zmsg, nodeid, flags) < 0)
         goto done;
-    if (json_str && !(o = Jfromstr (json_str))) {
-        errno = EPROTO;
+    if (flux_sendmsg (rpc->h, &zmsg) < 0)
         goto done;
-    }
-    if ((!o && out)) {
-        errno = EPROTO;
+    rc = 0;
+done:
+    zmsg_destroy (&zmsg);
+    return rc;
+}
+
+
+bool flux_rpc_check (flux_rpc_t rpc)
+{
+    if (rpc->oneway)
+        return false;
+    garbage_collect (rpc);
+    if (!rpc->rx_msg && rpc_recvmsg (rpc, true) < 0)
+        errno = 0;
+    if (rpc->rx_msg)
+        return true;
+    return false;
+}
+
+int flux_rpc_get (flux_rpc_t rpc, uint32_t *nodeid, const char **json_str)
+{
+    int rc = -1;
+
+    if (rpc->oneway) {
+        errno = EINVAL;
         goto done;
     }
-    if ((o && !out)) {
-        Jput (o);
-        errno = EPROTO;
+    garbage_collect (rpc);
+    if (!rpc->rx_msg && rpc_recvmsg (rpc, false) < 0)
         goto done;
+    rpc->rx_msg_consumed = true;
+    if (flux_response_decode (rpc->rx_msg, NULL, json_str) < 0)
+        goto done;
+    if (nodeid) {
+        uint32_t matchtag;
+        if (flux_msg_get_matchtag (rpc->rx_msg, &matchtag) < 0)
+            goto done;
+        *nodeid = lookup_nodeid (rpc, matchtag);
     }
-    if (out)
-        *out = o;
     rc = 0;
 done:
-    if (match.matchtag != FLUX_MATCHTAG_NONE)
-        flux_matchtag_free (h, match.matchtag, match.bsize);
-    zmsg_destroy (&zmsg);
     return rc;
 }
 
+/* N.B. if the user's 'then' callback doesn't call flux_rpc_get(),
+ * the callback will be called again (ad infinitum)
+ */
+static int rpc_cb (flux_t h, int type, zmsg_t **zmsg, void *arg)
+{
+    flux_rpc_t rpc = arg;
+
+    assert (rpc->then_cb != NULL);
+    do {
+        garbage_collect (rpc);
+        if (!rpc->rx_msg)
+            rpc_addmsg (rpc, zmsg);
+        rpc->then_cb (rpc, rpc->then_arg);
+    } while (*zmsg != NULL);
```

Comment:
I think it depends on the semantics we want for then.  If it is meant to consume the request, then it would be reasonable to discard that response whether get is called or not.   Is there a reasonable way to grab a handle to the request that the callback is being passed so it can be collected after the call if the call doesn't collect it?